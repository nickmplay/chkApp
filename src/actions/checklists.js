//checklist action generators

//add newChecklist object for dispatch
export const addChecklist = (newChecklist) => ({
  type: 'ADD_CHECKLIST',
  newChecklist
});

//add removeChecklist object for dispatch
export const removeChecklist = (checklistToRemoveId) => ({
  type: 'REMOVE_CHECKLIST',
  checklistToRemoveId
});