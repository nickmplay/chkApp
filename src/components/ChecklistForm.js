import React from 'react';

export class ChecklistForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      command: props.command,
      name: props.checklist ? props.checklist.name : '',
      items: props.checklist ? props.checklist.items.join(', ') : '',
      id: props.checklist ? props.checklist.id : ''
    };
  }
  onNameChange = (e) => {
    const name = e.target.value;
    this.setState(() => ({ name }));
  };
  onItemChange = (e) => {
    const newItem = e.target.value;
    this.setState(() => ({ items: newItem }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    //form validation
    if(this.state.name.length < 1 && this.state.items.length < 1){
      alert('Please enter a checklist name and items');
      return false;
    }
    
    if(this.state.name.length < 1){
      alert('Please enter a checklist name');
      return false;
    }

    if(this.state.items.length < 1){
      alert('Please enter some items for the checklist');
      return false;
    }

    // submit validated form
    this.props.onSubmit({
      name: this.state.name,
      items: this.state.items.split(', '),
      id: this.state.id
    });
  };

  render() {
    return (
      <div className="content-container">
        <form className="form" onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Checklist Name'
            value={this.state.name}
            onChange={this.onNameChange}
          />
          <textarea
            type='text'
            placeholder='Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3'
            value={this.state.items}
            onChange={this.onItemChange}
          />
          <button className="button">{this.state.command == "Add" ? "Add" : "Save"} Checklist</button>
        </form>
      </div>
    );
  }
}