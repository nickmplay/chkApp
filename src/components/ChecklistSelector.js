import React from 'react';
import { Link } from 'react-router-dom';

export const ChecklistSelector = (props) => {
  return (
    <div>
      {props.checklists.map(checklist => {
        return (
          <div className="content-container list-item" key={checklist.id}>
            <Link to={`/view/${checklist.id}`}>
              <p  key={checklist.id + 1}>{checklist.name}</p>
            </Link>
            <Link key={checklist.id + 2} to={`/edit/${checklist.id}`}>Edit</Link>
          </div>
        );
      })}
    </div>);
};