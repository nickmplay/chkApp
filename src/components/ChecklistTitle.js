import React from 'react';
import { countChecklists } from '../selectors/checklist';

export const ChecklistTitle = (props) => {
  const checklistsWord = props.checklists.length === 1 ? 'checklist' : 'checklists';

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">You have <span>{props.checklists.length}</span> {checklistsWord} to use:</h1>
        {/* <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div> */}
      </div>
    </div>
  );
};

export default ChecklistTitle;
