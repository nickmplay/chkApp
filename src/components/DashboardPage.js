import React from 'react';

const DashboardPage = () => (
  <div className="content-container">
    <h1>You have two checklists to use:</h1>
    <div>Checklist 1</div>
    <div>Checklist 2</div>
  </div>
);

export default DashboardPage;
