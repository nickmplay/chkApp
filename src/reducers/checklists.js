//checklist reducer function

const defaultChecklistState = [];

export default (state = defaultChecklistState, action) => {
  switch (action.type) {
    case 'ADD_CHECKLIST':
      return [...state, action.newChecklist];
    case 'REMOVE_CHECKLIST':
      return state.filter(({ id }) => id !== action.checklistToRemoveId);
    case 'EDIT_CHECKLIST':
      return state.map((e) => {
        if (e.id === action.checklistUpdates.id) {
          return {
            id: e.id,
            items: action.checklistUpdates.items
          };
        } else {
          return e; 
        }
      });
    default:
      return state;
  }
};
