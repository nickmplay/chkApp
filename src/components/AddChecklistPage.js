import React from 'react';
import { connect } from 'react-redux';
import { addChecklist } from '../actions/checklists';
import { startAddChecklist } from '../actions/firebase';
import { ChecklistForm } from './ChecklistForm';

export class AddChecklistPage extends React.Component {
  onSubmit = ({name, items}) => {
    this.props.startAddChecklist({
      name,
      items
    });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Checklist</h1>
          </div>
        </div>
        <ChecklistForm 
          command="Add"
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddChecklist: (checklist) => dispatch(startAddChecklist(checklist))
});

export default connect(undefined, mapDispatchToProps)(AddChecklistPage);
