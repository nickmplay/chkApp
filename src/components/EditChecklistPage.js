import React from 'react';
import { connect } from 'react-redux';
import { editChecklist } from '../actions/checklists';
import { ChecklistForm } from './ChecklistForm';

export class EditChecklistPage extends React.Component {
  onSubmit = (checklist) => {
    this.props.editChecklist(checklist);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit {this.props.checklist.name} checklist</h1>
          </div>
        </div>
        <ChecklistForm 
          command="Edit"
          checklist={this.props.checklist}
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  checklist: state.checklists.find( e => e.id.toString() === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editChecklist: (checklist) => dispatch(editChecklist(checklist))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditChecklistPage);
