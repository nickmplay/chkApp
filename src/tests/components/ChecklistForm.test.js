import React from 'react';
import { shallow } from 'enzyme';
import { ChecklistForm } from '../../components/ChecklistForm';
import checklists from '../fixtures/checklists';

//initial render
test('should render ChecklistForm correctly', () => {
  const wrapper = shallow(<ChecklistForm command="Add" />);
  expect(wrapper).toMatchSnapshot();
});

//input -> state tests
test('should set name if input changes', () => {
  const value = 'new name';
  const wrapper = shallow(<ChecklistForm command="Add" />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.state('name')).toBe(value);
});

test('should set items if textarea changes', () => {
  const value = 'first item, second item';
  const wrapper = shallow(<ChecklistForm command="Add" />);
  wrapper.find('textarea').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.state('items')).toBe(value);
});


//submit function tests - add
test('should handle onSubmit', () => {
  const onSubmit = jest.fn();
  const checklistToAdd = {
    name: checklists[1].name, 
    items: checklists[1].items,
    id: ''
  };
  const wrapper = shallow(<ChecklistForm
    command="Add"
    onSubmit={onSubmit}
    checklist={checklistToAdd}
  />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper).toMatchSnapshot();
  expect(onSubmit).toHaveBeenLastCalledWith(checklistToAdd);
});

test('should handle onSubmit manual', () => {
  const onSubmit = jest.fn();
  const checklistToAdd = {
    name: checklists[1].name, 
    items: checklists[1].items
  };
  const wrapper = shallow(<ChecklistForm
    command="Add"
    onSubmit={onSubmit}
  />);
  wrapper.find('input').at(0).simulate('change', {
    target: { value:checklistToAdd.name }
  });
  wrapper.find('textarea').at(0).simulate('change', {
    target: { value:checklistToAdd.items.join(', ') }
  });
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper).toMatchSnapshot();
  expect(onSubmit).toHaveBeenLastCalledWith({
    name: wrapper.state('name'),
    items: wrapper.state('items').split(', '),
    id: ''
  });
});

//submit function tests - edit
test('should handle onSubmit manual edit', () => {
  const onSubmit = jest.fn();
  const checklistToEdit = {
    name: checklists[1].name, 
    items: checklists[1].items,
    id: checklists[1].id
  };
  const newItemsStr = checklistToEdit.items.join(', ') + ', another item';
  const newName = checklistToEdit.name + ' updated';
  const wrapper = shallow(<ChecklistForm
    command="Edit"
    checklist={checklistToEdit}
    onSubmit={onSubmit}
  />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('input').at(0).simulate('change', {
    target: { value: newName }
  });
  wrapper.find('textarea').at(0).simulate('change', {
    target: { value:newItemsStr }
  });
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  });
  expect(wrapper).toMatchSnapshot();
  expect(onSubmit).toHaveBeenLastCalledWith({
    name: wrapper.state('name'),
    items: newItemsStr.split(', '),
    id: wrapper.state('id')
  });
});
