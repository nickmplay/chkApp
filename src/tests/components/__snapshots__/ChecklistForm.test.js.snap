// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle onSubmit 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value="car"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value="rucksack 1, rucksack 2"
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should handle onSubmit manual 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value="car"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value="rucksack 1, rucksack 2"
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should handle onSubmit manual edit 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value="car"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value="rucksack 1, rucksack 2"
    />
    <button
      className="button"
    >
      Save
       Checklist
    </button>
  </form>
</div>
`;

exports[`should handle onSubmit manual edit 2`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value="car updated"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value="rucksack 1, rucksack 2, another item"
    />
    <button
      className="button"
    >
      Save
       Checklist
    </button>
  </form>
</div>
`;

exports[`should not run onSubmit if validation fails - both 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should not run onSubmit if validation fails - items 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value="bed time"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should not run onSubmit if validation fails - name 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value="lock back door, lock front door, turn off lights"
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should render ChecklistForm correctly 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should set items if textarea changes 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value="first item, second item"
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;

exports[`should set name if input changes 1`] = `
<div
  className="content-container"
>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      placeholder="Checklist Name"
      type="text"
      value="new name"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add Items, seperating each with a comma. Eg Item 1, Item 2, Item 3"
      type="text"
      value=""
    />
    <button
      className="button"
    >
      Add
       Checklist
    </button>
  </form>
</div>
`;
